File: ai-radio-backend/src/services/ai/prompts.ts
/**
 * AI Prompt Templates
 * System prompts for generating engaging podcast conversations
 */

import { ScriptGenerationContext } from '../../types/podcast';

/**
 * System prompt for two-host podcast conversation generation
 */
export const PODCAST_SYSTEM_PROMPT = `You are a professional podcast script writer specializing in daily briefings. Your task is to create an engaging, natural conversation between two hosts for a personalized morning radio show.

HOST PERSONALITIES:
- Host 1 (Alex): Upbeat, energetic, and enthusiastic. Uses casual language and brings excitement to the conversation. Often asks questions and drives the discussion forward.
- Host 2 (Jordan): More analytical and thoughtful. Provides context and deeper insights. Balances Alex's energy with calm, clear explanations.

CONVERSATION GUIDELINES:
1. Natural Banter: The hosts should interact naturally, with back-and-forth dialogue, occasional jokes, and smooth transitions
2. Concise Summaries: Don't read entire emails or descriptions - summarize the key points conversationally
3. Time-Aware: Mention specific times naturally (e.g., "You've got that 2pm meeting coming up")
4. Engaging Transitions: Use smooth transitions between topics (emails → calendar → weather)
5. Personal Touch: Address the listener directly occasionally (e.g., "Looks like you have a busy afternoon ahead")
6. Target Duration: Aim for approximately 5 minutes of audio (~750-900 words total)

CONTENT STRUCTURE:
1. INTRO (15-20 seconds):
   - Warm greeting based on time of day
   - Brief overview of what's coming up
   - Set energetic, positive tone

2. EMAIL HIGHLIGHTS (90-120 seconds):
   - Summarize 3-5 most important emails
   - Group by theme when possible (e.g., "a few project updates", "couple of meeting requests")
   - Highlight action items naturally
   - Skip promotional/newsletter content unless specifically requested

3. CALENDAR OVERVIEW (60-90 seconds):
   - Start with immediate next event
   - Highlight 2-4 key meetings/events
   - Note any conflicts or tight schedules
   - Mention preparation needs if relevant

4. WEATHER (20-30 seconds, if included):
   - Current conditions and forecast
   - Practical advice (umbrella, jacket, etc.)
   - Keep it brief and relevant

5. TOPIC TEASERS (20-30 seconds, if topic previews provided):
   - Brief mentions of trending stories in user's followed topics
   - Encourage listener to explore topics for more details
   - Keep it engaging and leave them curious

6. OUTRO (15-20 seconds):
   - Quick recap or motivational note
   - Warm sign-off
   - Encourage a great day

TONE & STYLE:
- Conversational and friendly, not robotic
- Use contractions naturally ("you've" not "you have")
- Vary sentence structure and length
- Include occasional verbal markers ("So", "Alright", "Now", "By the way")
- Keep it professional but warm
- No overly formal language

WHAT TO AVOID:
- Don't read full email bodies or subjects verbatim
- Don't mention spam, promotions, or unimportant content
- Don't use technical jargon without context
- Don't make the conversation feel scripted
- Don't exceed 5 minutes of estimated speaking time

FORMAT YOUR RESPONSE as a JSON array of segments:
[
  {
    "speaker": "host1",
    "text": "Good morning! Welcome to your daily briefing for...",
    "type": "intro"
  },
  {
    "speaker": "host2",
    "text": "Thanks Alex! Let's dive into what's happening today...",
    "type": "intro"
  },
  ...
]

CRITICAL - Valid "type" values are ONLY these 7 options:
- "intro" - Opening greetings and overview
- "email" - Email summaries and discussion
- "calendar" - Calendar events and scheduling
- "news" - News topics discussion (use for general topics)
- "weather" - Weather information
- "teaser" - Topic teasers and previews for user's followed topics
- "outro" - Closing remarks and sign-off

Do NOT use any other type values. If unsure, use "email" for email content or "news" for general content.

Each segment should be a natural speaking turn, typically 1-3 sentences.`;

/**
 * Generate user-specific prompt with context
 */
// Map language codes to full names for the prompt
const LANGUAGE_NAMES: Record<string, string> = {
  en: 'English',
  es: 'Spanish',
  fr: 'French',
  de: 'German',
  it: 'Italian',
  pt: 'Portuguese',
  hi: 'Hindi',
  ja: 'Japanese',
  ko: 'Korean',
  zh: 'Chinese',
  ar: 'Arabic',
  ru: 'Russian',
  nl: 'Dutch',
  pl: 'Polish',
  tr: 'Turkish',
  vi: 'Vietnamese',
  th: 'Thai',
  id: 'Indonesian',
  ms: 'Malay',
  ta: 'Tamil',
  te: 'Telugu',
  bn: 'Bengali',
  mr: 'Marathi',
  gu: 'Gujarati',
  kn: 'Kannada',
  ml: 'Malayalam',
  pa: 'Punjabi',
};

export function generatePodcastPrompt(context: ScriptGenerationContext): string {
  const { user_name, date, time_of_day, content, preferences } = context;

  const greeting = time_of_day === 'morning'
    ? 'Good morning'
    : time_of_day === 'afternoon'
    ? 'Good afternoon'
    : 'Good evening';

  // Get language name from code, default to English
  const languageCode = preferences.language || 'en';
  const languageName = LANGUAGE_NAMES[languageCode] || 'English';

  console.log(`[generatePodcastPrompt] Language code: ${languageCode}, Language name: ${languageName}`);

  const languageInstruction = languageCode !== 'en'
    ? `\n\nIMPORTANT: Generate the ENTIRE script in ${languageName}. All dialogue, greetings, and content must be in ${languageName}. Do NOT use English except for proper nouns (names, places, etc.).\n`
    : '';

  let prompt = `Create a ${preferences.briefing_time} podcast script for ${user_name || 'the listener'} on ${date}.${languageInstruction}

CONTENT TO COVER:
`;

  // Email section
  if (preferences.include_email && content.emails.total_count > 0) {
    prompt += `
EMAILS (${content.emails.total_count} total, ${content.emails.important_count} important):
`;
    if (content.emails.highlights.length > 0) {
      prompt += `Key highlights:\n${content.emails.highlights.map((h, i) => `${i + 1}. ${h}`).join('\n')}\n`;
    }
    if (content.emails.action_items.length > 0) {
      prompt += `Action items:\n${content.emails.action_items.map((a, i) => `${i + 1}. ${a}`).join('\n')}\n`;
    }
  }

  // Calendar section
  if (preferences.include_calendar && content.calendar.total_events > 0) {
    prompt += `
CALENDAR (${content.calendar.total_events} events today):
`;
    if (content.calendar.next_event) {
      prompt += `Next up: ${content.calendar.next_event.title} - ${content.calendar.next_event.time_until}`;
      if (content.calendar.next_event.location) {
        prompt += ` at ${content.calendar.next_event.location}`;
      }
      prompt += '\n';
    }
    if (content.calendar.key_meetings.length > 0) {
      prompt += `Key meetings:\n${content.calendar.key_meetings.map((m, i) => `${i + 1}. ${m}`).join('\n')}\n`;
    }
  }

  // Weather section
  if (preferences.include_weather && content.weather) {
    const { weather } = content;
    prompt += `
WEATHER for ${weather.location}:
Current: ${weather.current.temperature}°F, ${weather.current.condition}
Forecast: High ${weather.forecast.high}°F, Low ${weather.forecast.low}°F - ${weather.forecast.condition}
`;
  }

  // Topic teasers section (for user's followed topics)
  if (content.topicTeasers && content.topicTeasers.length > 0) {
    prompt += `
TOPIC TEASERS (brief previews of user's followed topics - include near the end before outro):
`;
    content.topicTeasers.forEach((teaser, i) => {
      prompt += `${i + 1}. ${teaser.topicName}: "${teaser.headlines[0] || 'Latest updates available'}"\n`;
    });
    prompt += `\nMention these briefly (1-2 sentences each) to tease upcoming content and encourage the listener to explore these topics.\n`;
  }

  prompt += `
Remember: Create an engaging conversation between Alex (energetic) and Jordan (analytical). Keep it natural, concise, and around 5 minutes total. Return ONLY valid JSON.`;

  return prompt;
}

/**
 * Email summarization prompt
 */
export const EMAIL_SUMMARIZATION_PROMPT = `Analyze these emails and extract:
1. Key highlights (important information, updates, decisions)
2. Action items (tasks, requests, deadlines)
3. Priority level for each email

Focus on:
- Work-related content (projects, meetings, decisions)
- Time-sensitive items
- Action items requiring response

Ignore:
- Promotional emails
- Newsletters (unless explicitly work-related)
- Automated notifications
- Spam

Return a structured summary suitable for a morning briefing.`;

/**
 * Calendar summarization prompt
 */
export const CALENDAR_SUMMARIZATION_PROMPT = `Analyze today's calendar events and provide:
1. Next upcoming event with time until it starts
2. List of key meetings (3-5 most important)
3. Any scheduling conflicts or back-to-back meetings
4. Total busy time vs free time

Consider:
- Meeting importance based on attendees and titles
- Time blocks that might need preparation
- Tight schedules or conflicts

Format the summary conversationally for a morning briefing.`;

/**
 * Content filtering prompt
 */
export const CONTENT_FILTER_PROMPT = `Determine if this content should be included in a daily briefing.

Include if:
- Work-related and actionable
- Time-sensitive
- From important contacts
- Contains decisions or updates

Exclude if:
- Promotional or marketing
- Automated notifications (unless critical)
- Spam or newsletters
- Social media notifications

Return: { "include": boolean, "reason": string, "priority": "low" | "medium" | "high" }`;

// TODO: Implement prompt optimization based on user feedback
// TODO: Add A/B testing for different prompt variations
// TODO: Create persona customization based on user preferences
// TODO: Add support for multiple languages
// TODO: Implement dynamic prompt adjustment based on content volume

File: audexa-radio/orchestrator/prompts.py
"""
Radio-format prompt templates for Claude script generation.
Adapted from ai-radio-backend prompts.ts with radio-specific instructions.
"""

from datetime import datetime, timezone

RADIO_SYSTEM_PROMPT = """You are a professional radio script writer for Audexa Radio, a 24/7 AI-powered news and talk station. Create engaging, natural conversation between two hosts.

HOST PERSONALITIES:
- Host 1 (Alex): Upbeat, energetic, and enthusiastic. Uses casual language, brings excitement, asks questions, and drives the discussion forward. Think of a morning radio host.
- Host 2 (Jordan): More analytical and thoughtful. Provides context and deeper insights. Balances Alex's energy with calm, clear explanations.

RADIO GUIDELINES:
1. This is LIVE RADIO — reference "listeners tuning in", the time of day, and the station name "Audexa Radio"
2. Natural banter with back-and-forth dialogue, occasional humor, and smooth transitions
3. Conversational and warm — use contractions, verbal markers ("So", "Alright", "Now", "Here's the thing")
4. Summarize stories concisely — don't read full articles. Hit the key points conversationally
5. End segments with teases: "Stick around, we've got more coming up after this track" or "Don't go anywhere"
6. Vary energy — not everything needs to be high energy. Deep dives can be more reflective

FORMAT YOUR RESPONSE as a JSON array of segments:
[
  {"speaker": "host1", "text": "...", "type": "intro"},
  {"speaker": "host2", "text": "...", "type": "headlines"},
  ...
]

Valid segment types: "intro", "headlines", "deep_dive", "listener_request", "transition", "outro"
Each segment should be a natural speaking turn, typically 1-3 sentences."""


def build_headlines_prompt(
    topic_name: str, stories_text: str, prompt_context: str,
    language: str = "en", region: str = "us",
) -> str:
    """Build prompt for a 2-3 minute headlines segment."""
    now = datetime.now(timezone.utc)
    time_of_day = _get_time_of_day(now.hour)
    lang_note = f"\nIMPORTANT: Write the entire script in {_language_name(language)}. Tailor references and context for {_region_name(region)} listeners." if language != "en" or region != "us" else ""

    return f"""Generate a HEADLINES segment for Audexa Radio.

Topic: {topic_name}
Time: {time_of_day} ({now.strftime('%I:%M %p UTC')})
Region: {_region_name(region)}
Duration target: 2-3 minutes (~350-500 words total across all segments){lang_note}

Context: {prompt_context}

Cover 4-5 of these stories as quick headlines with brief commentary:

{stories_text}

Start with a brief intro referencing the time and topic. Cover each story in 2-3 speaking turns. End with a tease for what's coming next. Return ONLY valid JSON."""


def build_deep_dive_prompt(
    topic_name: str, stories_text: str, prompt_context: str,
    language: str = "en", region: str = "us",
) -> str:
    """Build prompt for a 3-5 minute deep dive segment."""
    now = datetime.now(timezone.utc)
    lang_note = f"\nIMPORTANT: Write the entire script in {_language_name(language)}. Tailor references and context for {_region_name(region)} listeners." if language != "en" or region != "us" else ""

    return f"""Generate a DEEP DIVE segment for Audexa Radio.

Topic: {topic_name}
Region: {_region_name(region)}
Duration target: 3-5 minutes (~500-750 words total across all segments){lang_note}

Context: {prompt_context}

Pick the 2-3 most interesting stories below and go deeper. Discuss implications, context, and what it means for listeners:

{stories_text}

The hosts should have a real discussion — Alex asks probing questions, Jordan provides analysis. End with "we'll be right back after this track." Return ONLY valid JSON."""


def build_listener_request_prompt(
    topic_text: str,
    language: str = "en", region: str = "us",
) -> str:
    """Build prompt for a listener-requested topic segment."""
    lang_note = f"\nIMPORTANT: Write the entire script in {_language_name(language)}." if language != "en" else ""

    return f"""Generate a LISTENER REQUEST segment for Audexa Radio.

A listener just called in and requested we cover: "{topic_text}"{lang_note}

Duration target: 2-4 minutes (~350-600 words total)

Start with Alex saying something like "We just got a call from a listener who wants to hear about {topic_text}" — make it feel special and acknowledged.

Then have both hosts discuss the topic naturally. Use your knowledge to cover it well. If it's a current event, provide context and implications. If it's a general topic, make it interesting and informative.

End with thanking the listener and a transition to the next segment. Return ONLY valid JSON."""


def build_transition_prompt() -> str:
    """Build a short DJ transition between segments."""
    now = datetime.now(timezone.utc)
    time_of_day = _get_time_of_day(now.hour)

    return f"""Generate a very short TRANSITION for Audexa Radio.

Time: {time_of_day}

Just 2-3 speaking turns (Alex and Jordan). Something like:
- Welcome back from the song
- Brief mention of what's coming up
- Keep it under 30 seconds of speaking time (~50-80 words)

Return ONLY valid JSON."""


def format_stories_for_prompt(stories: list) -> str:
    """Format aggregated stories into a text block for the prompt."""
    lines = []
    for i, story in enumerate(stories[:10], 1):
        line = f"{i}. [{story.source}] {story.title}"
        if story.summary:
            line += f"\n   {story.summary[:200]}"
        if story.score:
            line += f" (score: {story.score})"
        lines.append(line)
    return "\n\n".join(lines)


def _language_name(code: str) -> str:
    names = {
        "en": "English", "de": "German", "fr": "French", "es": "Spanish",
        "pt": "Portuguese", "ja": "Japanese", "ko": "Korean", "hi": "Hindi",
        "ar": "Arabic", "zh": "Mandarin Chinese",
    }
    return names.get(code, code)


def _region_name(code: str) -> str:
    names = {
        "us": "United States", "uk": "United Kingdom", "in": "India",
        "au": "Australia", "ca": "Canada", "de": "Germany", "fr": "France",
        "es": "Spain", "br": "Brazil", "jp": "Japan", "kr": "South Korea",
        "sg": "Singapore", "ae": "UAE", "za": "South Africa",
    }
    return names.get(code, code.upper())


def _get_time_of_day(hour_utc: int) -> str:
    # Rough mapping — adjust for listener timezone later
    if 5 <= hour_utc < 12:
        return "morning"
    elif 12 <= hour_utc < 17:
        return "afternoon"
    elif 17 <= hour_utc < 21:
        return "evening"
    else:
        return "late night"
